### ISSUE TITLE
Make set_axisbelow work to be behind even for graphs with alpha

### ISSUE THREAD

**Issue by snowzky** [matplotlib/matplotlib#18831]
I request a feature change of how set_axisbelow works. One that still makes grid lines appear below a plot, even when that plot has alpha.

There are many kinds of data science where you would still like the help of grid lines, but where the plot can become messy if you see the grid lines OVER your data. But at the same time, there can be times where you may have overlapping data points, so you need alpha. The combined case therefor also happens.

I would like to be able to make a plot where my bars, lines, markers or whatever uses the alpha channel, but that the grid lines are still behind those markers/lines/bars/whatever.
Basically make set_axisbelow take an argument like (account_for_alpha=False, True) or some other smart naming.

**Comment by jklymak** [comment]
Can you provide a reproducible example of the poor behaviour for specificity?

**Comment by snowzky** [comment]
> Can you provide a reproducible example of the poor behaviour for specificity?

Modified [Stacked bar chart from Matplotlibs ](https://matplotlib.org/3.3.2/gallery/lines_bars_and_markers/bar_stacked.html#sphx-glr-gallery-lines-bars-and-markers-bar-stacked-py) with an alpha for the bar charts, and with enabling the yaxis-major-grid lines and the set_axesbelow.

`import matplotlib.pyplot as plt

labels = ['G1', 'G2', 'G3', 'G4', 'G5']
men_means = [20, 35, 30, 35, 27]
women_means = [25, 32, 34, 20, 25]
men_std = [2, 3, 4, 1, 2]
women_std = [3, 5, 2, 3, 3]
width = 0.35       # the width of the bars: can also be len(x) sequence

fig, ax = plt.subplots()

ax.bar(labels, men_means, width, yerr=men_std, label='Men',alpha=0.6)
ax.bar(labels, women_means, width, yerr=women_std, bottom=men_means,
       label='Women',alpha=0.6)


ax.set_ylabel('Scores')
ax.set_title('Scores by group and gender')
ax.legend()
ax.yaxis.grid(True, which='major')
ax.set_axisbelow(True)

plt.show()`

Simplest example I could quickly find. Maybe this case doesn't look so bad to some people (though I personally think it looks bad with the lines cutting across the bars), and in this example, no, you don't need alpha for the bars, as they don't overlap.
But this shows how alpha interact with the set_axisbelow (or rather, how they don't interact.)

![matplotlib_alpha_setaxisbelow_feature_request](https://user-images.githubusercontent.com/40359391/97428438-460a7f00-1916-11eb-84b2-0a850806b210.png)

**Comment by jklymak** [comment]
Alpha makes the bars see through. The grid lines are under the bars, but you are making the bars transparent so they show up.  I dont think there is anything we can, or want, to do about that.

**Comment by tacaswell** [comment]
Jody is correct, this is the expected and desired behavior.  If you make the grid lines thicker and colored via

```python
ax.yaxis.grid(True, which='major', lw=15, color='r')
```

you get 
![so](https://user-images.githubusercontent.com/199813/97490669-89c1b080-1937-11eb-9b00-9183ef41fa7b.png) which makes it clear that the blue bars transparent and blended into what ever is behind them.

**Comment by snowzky** [comment]
I don't understand your answers. 

I _know_ this is expected behaviour. I want a _new_ behaviour. That's why I labelled this feature request, not bug.
I am asking for an enhancement of the behaviour, where you can get both the behaviour as it is now, and also that the grid lines don't show "through" plots even when they have transparency.

**Comment by jklymak** [comment]
Grid lines are too low level to consider adding a lot of magic for them to check the other artists in the figure and hide portions of themselves somehow.  You can, of course, manually draw lines yourself in the appropriate gaps if you want, but it will get ugly in a hurry for general cases...
